### DISCUSSION TITLE
Using pymor for Model reduction of FEA models

### DISCUSSION THREAD

**Discussion by clausob** [pymor/pymor#1922] [Q&A]
Hello,

I should start with a short introduction: I am a mechanical engineer, working in CAE-simulation (especially FEA) for some 
time, doing this just out of interest.

I am interested in FMU/FMI models (https://fmi-standard.org/), thinking about to se if it is possible to add a free 
FEA software to that environment. I am thinking about a workflow:

FEA software: currently calculix, generates stiffness and mass-matrix from discretization
MOR:             pymor (well, that's the reason I am asking): generates reduced model from 
                      FEA-Matrices
FMU/FMI:      An SDK, not really thought about because I am stuck with MOR

My approch regarding MOR: I read in the FEA-matrices, generate the ROM model and export it. here I am stuck.
As I am still quite new to pymor and also to Python, my fist question would be if I am doing something wrong,
havin the wrong apprioach, etc.

Starting point: mass and stiffnes is stored in files, read in and then ... but let's take a look at my comments
BTW: Model is a simple beam, 1st Eigenfrequenzy 186 Hz, other also reasonable, so no special model

Would be very grateful if you could help!!! Script below.

Claus

```python
import scipy.io as sio

from scipy.sparse import lil_matrix
from scipy.sparse import identity

import numpy as np

#import sys
#if len(sys.argv) != 2:
#    print ("Usage: python ", sys.argv[0], " Calculixbasename")
#    quit()
#c_filename = sys.argv[1] + "_C.mtx"
#m_filename = sys.argv[1] + "_M.mtx"

m_filename = "FEMMeshNetgen" + "_M.mtx"
c_filename = "FEMMeshNetgen" + "_C.mtx"


C_mat=lil_matrix(sio.mmread(c_filename))
M_mat=lil_matrix(sio.mmread(m_filename))


mat_dim = C_mat.get_shape()[1]
 
# #
# # check symmetry
# #
# v_one = np.ones(mat_dim)
# c_sym_err = (C_mat - C_mat.T) * (C_mat - C_mat.T) 
# print(c_sym_err.dot(v_one).dot(v_one))
# quit()


D_mat = lil_matrix((mat_dim,mat_dim))

num_inputs = 1
B_mat = lil_matrix((mat_dim,num_inputs))
B_mat[942-1,0] = 1.0

num_outputs = 1 
Cp = lil_matrix((num_outputs,mat_dim))
Cp[0,3-1] = 1.0


from pymor.models.iosys import SecondOrderModel
# Full Model
som = SecondOrderModel.from_matrices(M_mat, D_mat, C_mat, B_mat, Cp, Cv=None, D=None, 
    sampling_time=0, state_id='STATE', solver_options=None,  error_estimator=None, visualizer=None, name=None)
# cont_time=True,
#estimator=None,



#from  pymor.reductors.sobt import SOBTReductor
#sobt = SOBTReductor(som)
#rom = sobt.reduce(20)
#
# 00:12 solve_lyap_lrcf: Relative residual at step 1: 1.75895e+03
# 00:25 solve_lyap_lrcf: Relative residual at step 2: 3.86281e+03
# 00:25 gram_schmidt: Orthonormalizing vector 1 again
# 00:39 solve_lyap_lrcf: Relative residual at step 3: 5.00625e+03
# 00:52 solve_lyap_lrcf: Relative residual at step 4: 4.53603e+03
# 00:52 gram_schmidt: Orthonormalizing vector 1 again
# 00:52 gram_schmidt: Orthonormalizing vector 2 again
# 00:52 gram_schmidt: Orthonormalizing vector 3 again
# 01:06 solve_lyap_lrcf: Relative residual at step 5: 4.03348e+03
# 01:32 solve_lyap_lrcf: Relative residual at step 7: 3.68578e+03
# .....
# no convergence


# ------------------------------------------------------------------------

#from  pymor.reductors.sobt import SOBTfvReductor
#sobt = SOBTfvReductor(som)
#rom = sobt.reduce(20)
#
# same as SOBTReductor
#
# ------------------------------------------------------------------------


#from pymor.reductors.sor_irka import SORIRKAReductor
#sobt = SORIRKAReductor(som)
#rom = sobt.reduce(20)
#
# 00:00 SORIRKAReductor: Generating initial interpolation data
# 00:00 SORIRKAReductor: Starting SOR-IRKA
#
# interrupted after some time

# ------------------------------------------------------------------------


# no change .....
#rom = sobt.reduce(20,projection='orth',tol=1e-2,num_prev=5)

# ------------------------------------------------------------------------


#lti=som.to_lti()
#from pymor.reductors.bt import BTReductor
#bt=BTReductor(lti)
#rom = bt.reduce(20)
#
#
# same as SOBTReductor, which makes sense
#

# ------------------------------------------------------------------------


#lti=som.to_lti()
#from pymor.reductors.bt import FDBTReductor
#bt=FDBTReductor(lti)
#rom = bt.reduce(20)
#
#00:00 |WARNING|LTIModel: Converting operator A to a NumPy array.
#00:00 |WARNING|LTIModel: Converting operator E to a NumPy array.
#
# converting sparse to dense matrices, does not work for FEA at reasonable number of DOF 
#

# ------------------------------------------------------------------------


#lti=som.to_lti()
#from pymor.reductors.bt import LQGBTReductor
#bt=LQGBTReductor(lti)
#rom = bt.reduce(20)
#
#
# same as SOBTReductor
#



print(rom)



quit()
```

**Comment by pmli** [comment]
Hi @clausob. (I edited your comment to use a Markdown code block.)

Your approach so far (exporting matrices from a PDE solver and importing them in pyMOR) is completely reasonable. I'm not familiar with FMU/FMI, but that could be something that pyMOR might support.

Looking at your code, it seems to me that the damping matrix (`D_mat`) is zero. This would mean that the system you are considering is not asymptotically stable, but many methods in pyMOR assume the opposite. Actually, a second-order model with positive definite mass and stiffness matrices and without damping has all its poles on the imaginary axis (if I'm not mistaken), which is very difficult to approximate, especially if you are interested in a large spectrum of frequencies.

If you really want to approximate a system without damping, you can try using modal truncation (`MTReductor`), which will try to find a few dominant poles. If you are interested in a particular range of frequencies, there are approaches for that (not all available in pyMOR, at least so far).

**Comment by clausob** [comment]
Hi Petar,

thanks a lot for your helpful response. A short report of work in progress:

As a first, fast approach I have used some structural damping (i.e. D_mat = k * C_mat, playing with k).
Convergency better, but without changing the the tolerance for convergency criterium, no result 
achievable. With loose tolerances, result still  far away from exact result and FOM-result. Still trying
out variants.

Also tried MTReductor (without damping), but no convergency yet.

As I said, it is work in progress. But I am still under the impression that a certain guidance or "Best practice"how the methods
implemented should be applied to my kind of problem is missing (to my knowledge). Any links of help would
be very much appreciated!

Best regards,

Claus

**Reply by pmli** [reply]
It is tricky to debug this without access to the model you are using. Are you able to share the matrices you generated? If not, could you say what is the system order (`som.order`) and the number of steps taken by `solve_lyap_lrcf` and `MTReductor` (and the reported residues)?

So far, we do not have a guide for selecting methods (there is an open issue #1458). Mostly, the choice of the method depends on the type of model you are considering (linear/nonlinear, first-order/second-order, parametric/non-parametric, ...). For non-parametric `SecondOrderModels`, there are variants of balanced truncation that preserve the second-order structure (e.g., `SOBT` and `SOBTfv` you used), as well as interpolatory and $\mathcal{H}_2$-based methods (`SOBHIReductor` and `SORIRKAReductor`). If you are not concerned about preserving the second-order structure, there are more general methods for `LTIModels` (e.g., `BTReductor` and `IRKAReductor`).

The methods can be further distinguished by how they measure error: $\mathcal{H}_\infty$ norm for balanced truncation, $\mathcal{H}_2$ norm for IRKA, local error for interpolation, matching of poles and residues for modal truncation.

Of course, this is incomplete, as there are many more methods that are not implemented in pyMOR (different system structures, different error measures, ...).

**Reply by clausob** [reply]
Hi Petar,

no, there are no issues sharing the model and matrices. You can find them here:

https://nestnet.homelinuxserver.org/nextcloud/s/p3wK3Mw87iBA4yN

I cleaned things a bit up, added a Readme.txt to make the meaning of the files transparent. I hope that it is complete.
If you need additional information (FE-results, ....?), please just let me know.

Regarding your comments on the "target" and its structure: preserving second order structure would be nice to have,
but if the physical nature that we are having a second order problem is reflected by the introduction of the first derivatives
of the solution (which I thought is always the case for the reduction methods), well it's ok (that's the reason I was not 
surprised when I got the same convergence results by using som_to_lti()+BTRecuctor or SOBTRreductor).

I would very much appreciate if you could take look at the problem. I still think that there might be a quite stupid mistake buried somewhere.

Thanks,

Claus

**Reply by pmli** [reply]
Great, thanks.

I could load both `M_mat` and `C_mat`, and both are indeed sparse symmetric matrices of size $10,\\!986 \times 10,\\!986$. Then I ran `eigsh` from SciPy (with option `which='BE'`) and got that the two smallest eigenvalues of `M_mat` are $-2.11 \cdot 10^{-10}$ and $1.54 \cdot 10^{-13}$, and the largest is $1.39 \cdot 10^{-8}$. This would suggest that `M_mat` is not positive definite, which is what might be causing issues. Could this be an issue in the PDE discretization?

> Regarding your comments on the "target" and its structure: preserving second order structure would be nice to have, but if the physical nature that we are having a second order problem is reflected by the introduction of the first derivatives of the solution (which I thought is always the case for the reduction methods), well it's ok (that's the reason I was not surprised when I got the same convergence results by using som_to_lti()+BTRecuctor or SOBTRreductor).

The main computation in `BTReductor` and `SOBTReductor` is the same: computing the controllability and observability Gramians (where an iterative, low-rank solver is used for larger models). The difference is what they do with the Gramians. The result should be that SOBT gives a (slightly) less accurate reduced-order model compared to BT (because SOBT is restricted to the second-order model structure).

**Reply by clausob** [reply]
Thanks a lot for your investigations (and findings, I have to admit). I think you found (the first?) of what I called 
buried stupid mistakes: I made a mistake doing the data conversion Calculix-->MatrixMarket, fixed it only for the C_mat..........

Update of the M_mat: https://nestnet.homelinuxserver.org/nextcloud/s/mWy9HrK4pNpAiBx

**Reply by drittelhacker** [reply]
Claus, 
I have done some tests with your latest model. Compared to your `pymor_test1.py` I did the following changes:

-  `D_mat = C_mat * 1e-6` your damping was much to strong to see any resonances. 
-  `som = SecondOrderModel.from_matrices(M_mat.tocsc(), D_mat.tocsc(), C_mat.tocsc(), B_mat.toarray(), Cp.toarray(), Cv=None, D=None, sampling_time=0, state_id='STATE', solver_options={ 'lyap_lrcf': {'type': 'lradi', 'tol': 1e-8}}, error_estimator=None, visualizer=None, name=None)` the lil_matrices completely ruined the performance, so I converted the square matrices to CSC format and the tall and skinny ones to dense arrays. I started of with lradi tol 1e-12 but got reasonable results with 1e-8 already.
- `rom = sobt.reduce(20)` the last thing you had was order 1 which was sufficient in the over-damped case but needed an increase for the model that actually showed some action

Finally, I added some evaluation: 
```python
import matplotlib.pyplot as plt
err = som -rom
w = (1e0, 1e8)
fig, axs = plt.subplots()
bode = som.transfer_function.mag_plot(w, ax=axs, Hz=True)
bode = err.transfer_function.mag_plot(w, ax=axs, Hz=True)
fig.savefig('ERR_mag.png')
```

And here is the result: (SOM in blue and error for the order 20 model in orange)
![ERR_mag](https://user-images.githubusercontent.com/29063401/213750149-7df08b2b-8364-41ff-980f-1a49fae69345.png)

The reduction results look reasonable to me, but you may want to revisit your original model export, as I do not see the first resonance at 186Hz.

Moreover, I have tried a different type of Rayleigh damping: `D_mat = M_mat * 1e-6` However, that shows lots of high frequency noise in the magnitude plot that seems rather unphysical. On the other hand, I can confirm that, then, convergence breaks down.
I've also tried to verify the results in another LR-ADI implementation outside python, but got things like NAN in the original transfer function already, which again indicates that something might be broken in your system matrices.

**Reply by clausob** [reply]
Jens,

thanks a lot for your hints and the evaluation.

I suspected as well that the original matrices are somewhat "damaged" and thought that the conversion from 
calculix mass & stiffness into MatrixMarket files are the reason (like in the first attempt). As the calculix format
and the MatrixMarket-Format are very close to each other (basically only the header and the index is 
different for symmetric matrices, if important), I wrote a simple sh/awk converter.

Well, with these matrices, I got the exact same plot as you did and in addition, when I tried to solve for 
eigenvalues with octave, same as with your approach, NaN......

I agree, i think I have to check the original output of calculix.

Thanks, 

Claus

**Reply by clausob** [reply]
Jens,

I checked the mass and stiffness matrices, 
- got the statuc response $C*u=f$, $u$ checked for POIs, correct
- $\sum_{ij} M_{ij} = 3 * Mass of structure$, which is correct for a 3D-problem
- generalized Eigenvalue-problem with C and M delivers correct Eigenvalues (at least the first)
Thus I am quite happy with C and M.

One question: in your plot, do we see on the horizontal axis the Frequency $f$ ("physical frequency")  or the Eigen-Frequency 
$\omega = 2 \pi f = \sqrt{\lambda}$?

If it is $\omega$, it would be quite OK!

Thanks,

Claus

**Comment by clausob** [comment]
PS: and ... well I got a typo in my first post, it's $f=168 \ Hz$ not 186, i.e. $\omega=1060 \ 1/s$, which would match quite well your plot
sorry.....

**Reply by drittelhacker** [reply]
As mentioned in my post above, I called `mag_plot` with `Hz=true` so this should be the real frequencies in Hertz, and the resonance peak should actually appear at 168. I will try to crosscheck in other implementations, to make sure there is no scaling bug `mag_plot`

**Reply by drittelhacker** [reply]
Claus,
I finally found the time to try this in the latest Matlab `bodemag` for `mechss`. Also, there, I see the first eigen-frequency at 5.22e+3 rad/s, i.e. about 831Hz.

**Reply by clausob** [reply]
Jens,

I have to apologize that I did not follow up on your remarks, thasnks a lot for your comments and work . The last weeks
have been a bit stressful, so work on this topic from my side was stopped.

I think that the approach and the usage is relatively clear, I think I will set up a new model from scratch to 
avoid inconsistencies.

Will come back with results then... 

Best regards,

Claus
